app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -71,4 +97,132 @@ object MessageController {
+                val pageEnd = min(pageStart + pageSize, uids.count())
+                val page = uids.subList(pageStart, pageEnd)
+                ApiRepository.getMessagesByUids(mailboxUuid, folder.id, page).data?.messages?.let { messages ->
+                    FolderController.updateFolder(folder.id) { folder ->
+                        folder.threads += messages.map { it.toThread(mailboxUuid) }.toRealmList()
+                    }
+                }
+                // TODO: Do we want a delay between each call, to not get blocked by the API?
+                delay(500L)
+                pageStart += pageSize
+            }
+        }
+    }
+
+    private fun deleteMessages(uids: List<String>, realm: MutableRealm) {
+        if (uids.isNotEmpty()) {
+            realm.deleteMessages(getMessages(uids, realm))
+            realm.delete(ThreadController.getThreads(uids, realm))
+        }
+    }
+
+    private fun updateMessages(messageFlags: List<MessageFlags>, mailboxUuid: String, folderId: String, realm: MutableRealm) {
+        messageFlags.forEach {
+            val uid = it.shortUid.toLongUid(folderId)
+            updateMessage(uid, realm) { message ->
+                message.answered = it.answered
+                message.isFavorite = it.isFavorite
+                message.forwarded = it.forwarded
+                message.scheduled = it.scheduled
+                message.seen = it.seen
+
+                ThreadController.upsertThread(message.toThread(mailboxUuid), realm)
+            }
+        }
+    }
+
+    private fun threeMonthsAgo(): String = SimpleDateFormat("yyyyMMdd", Locale.ROOT).format(Date().monthsAgo(3))
+
+    private fun String.toLongUid(folderId: String) = "${this}@${folderId}"
+
+    private fun String.toShortUid(): String = substringBefore('@')
+
+    private fun getMessagesUids(mailboxUuid: String, folderId: String): MessagesUids {
+        return MessagesUids().apply {
+            with(ApiRepository.getMessagesUids(mailboxUuid, folderId, threeMonthsAgo())) {
+                if (isSuccess()) with(data!!) {
+                    this@apply.addedShortUids = addedShortUids
+                    this@apply.cursor = cursor
+                }
+            }
+        }
+    }
+
+    private fun getMessagesUidsDelta(mailboxUuid: String, folderId: String, previousCursor: String): MessagesUids {
+        return MessagesUids().apply {
+            with(ApiRepository.getMessagesUidsDelta(mailboxUuid, folderId, previousCursor)) {
+                if (isSuccess()) with(data!!) {
+                    this@apply.addedShortUids = addedShortUids
+                    this@apply.deletedUids = deletedShortUids.map { it.toLongUid(folderId) }
+                    this@apply.updatedMessages = updatedMessages
+                    this@apply.cursor = cursor
+                }
+            }
+        }
+    }
+
+    private fun getUniquesUidsInReverse(folder: Folder?, remoteUids: List<String>): List<String> {
```

Comment:
Folder is never null. Why did you put a null type on it?

Reply:
Fixed.